[issue #914] _P053_PMSx003 Dust Sensor looses sync

TD-er opened the pull request:
- Use peek instead of read, to make sure a complete packet is in the buffer.
- Flush buffer after processing a packet.
- Increased software serial buffer to the size of 3 packets.
- Peek (+ read garbage) 10 times per second to get the buffer in sync to start with start byte of packet.
- Decent delete + clean up of software serial allocation when changing settings (prevent memory leak)
- Move log String allocation to inside scope of function, not permanent allocated.
- Split into separate functions to make it better readable.

TD-er committed:
[issue #914] _P053_PMSx003 Dust Sensor looses sync

- Use peek instead of read, to make sure a complete packet is in the buffer.
- Flush buffer after processing a packet.
- Increased software serial buffer to the size of 3 packets.
- Peek (+ read garbage) 10 times per second to get the buffer in sync to start with start byte of packet.
- Decent delete + clean up of software serial allocation when changing settings (prevent memory leak)
- Move log String allocation to inside scope of function, not permanent allocated.
- Split into separate functions to make it better readable.

uzi18 commented:
It is better to check also if dummy (marker) is exactly 0x424D,  if not we don't need to try calculate checksum.

Looks like problem was because always take 2 bytes from buffer and datagram has even bytes so it is impossible to find sync if miss one byte.

TD-er commented:
That's a good suggestion.  (it was past 2am)
Will try that later today.

TD-er committed:
[issue #914] Check first 2 bytes of packet

TD-er committed:
[Build error fix] Must-keep-focus-when-pressing-buttons...

Quickly commit + merge is never a good idea when some visitors arrive...
Fixed now.

micropet commented:
OK, test begins.

TD-er commented:
@micropet OK to merge it?

micropet commented:
Unfortunately, no. There were about 60 minutes of data delivered.
Now nothing is coming out.

TD-er commented:
But it is not worse than before?
When I merge it, then others can also test. Just to get an idea what's going on. (don't have one of those sensors myself, for the next few weeks)

micropet commented:
Yes you are right. This version also runs longer than the old version.

micropet commented:
[serialReadPMValue.zip](https://github.com/letscontrolit/ESPEasy/files/1772961/serialReadPMValue.zip)

I've been running this version on an ESP32 and STMF4 board for many months now.

She has never failed. Maybe it gives you an idea.

  #include "serialReadPMValue.h"
  uint16_t p01  = 0;         //define PM1.0 value of the air detector module
  uint16_t p25  = 0;         //define PM2.5 
  uint16_t p10  = 0;         //define PM10
  #define receiveDatIndex 32
  uint8_t receiveDat[receiveDatIndex]; 

  int length = serialRead(Serial1,receiveDat,receiveDatIndex,5); //change the serial port:Serial1,Serial2..
  int checkSum=checkValue(receiveDat,receiveDatIndex);
 
  if(length&&checkSum)
  {
    p01=transmitPM01(receiveDat);   // count PM1.0 value of the air detector module
    p25=transmitPM2_5(receiveDat);  // count PM2.5 value of the air detector module
    p10=transmitPM10(receiveDat);   // count PM10 value of the air detector module
  }
